### ISSUE TITLE
Reuse blocks from versioning directories

### ISSUE THREAD

**Issue by milefork** [syncthing/syncthing#5279]
Hi,

I today I "broke" my pic lib. I've 2 Servers with syncthing mirroring it over 2 different places. That was very stable but today I added my workstation with the same lib (but with right file attributes because I broke them on the servers due to rsync...). So to preserve my attributes I set the Workstation folder to send only.
While the workstation was initial scanning there was a button to override changes or smth. I pressed it and the syncthing from my workstation updated the global folder status to its local and continued scanning.
The servers, holding the whole lib started after that to remove the files not listed in the status... -> legitimate.

**BUT** I had set the servers to do simple file versioning. So all of the files are still available on the servers but they don't use them (or parts of it) and instead downloading it again over DSL/Cable....

tl;dr

Make use of blobs/files available from .stversions folder.

**Comment by AudriusButkevicius** [comment]
Yeah, I guess we should, though there are workarounds for this, search the forums.

**Comment by AudriusButkevicius** [comment]
I also think this is a duplicate, did you check existing issues before opening this one?

**Comment by milefork** [comment]
Okay I'll search the forums but it would be nice to see this implemented.
Yes I searched the issues but I hadn't found one.

**Comment by wweich** [comment]
dup of #1909 ?

**Comment by uok** [comment]
I still think this is a good idea and it saves a lot of bandwidth on slower connections. Please reopen #1909

**Comment by AudriusButkevicius** [comment]
Didn't realise it was closed. I'll reopen this one.

**Comment by milefork** [comment]
So I read through #1909 in which this part is covered as scenario 2.
As I see they are drifting a little bit away from the point that it should only be used as a source for same files and going to wan't it synced too.

That was not my intention here! My idea was to only let syncthing use the binary from .stversions as a source for files to be synced(downloaded).

It seems that you can trick syncthing to do that already with this here https://github.com/syncthing/syncthing/issues/1909#issuecomment-215340063

But it would be nice if it could do that out of the box. Whith that you have no corruption or something at all, just saving bandwidth as @uok  says.

**Comment by fragtion** [comment]
It seems there's an increasing number of users who would desire to cache .stversions on demand, for specific use cases. I'm also interested in doing this.

I tried the suggestion by @uok at #1909. While it does work perfectly at first, after a while the .stfolder marker is removed (I'm guessing because the .stversions is as a subfolder of an existing "Folder" repo, so that instance considers this .stfolder as alien)

Does anyone know how to work around this behaviour? I tried to add .stfolder and .stversions to the ignore list of the parent Folder, but this didn't seem to help.   Ideally, there should be an (advanced?) option for each folder allowing to toggle whether the .stversions should be indexed or not. Perhaps this could be a checkbox close to where the user can already select the type of archiving scheme. Your thoughts, peeps?

**Comment by calmh** [comment]
That's probably the versioner being configured to clean out files older than something, and also empty directories. You could set the marker name to point to something that is guaranteed to exist, or keep the marker alive yourself somehow.

**Comment by uok** [comment]
I have the following file structure
```` 
Master folder
 +--Folder A
 +--Folder B
 +--Folder C
```` 
Each subfolder (Folder A, Folder B, etc.) corresponds to a shared folder in Syncthing. All I did was to add "Master folder" as an unshared Syncthing folder. This works great and now looks like this:
```` 
Master folder
 .stfolder
 +--Folder A
  .stfolder
  .stversions
  <Files>
 +--Folder B
  .stfolder
  .stversions
  <Files>
 +--Folder C
  .stfolder
  .stversions
  <Files>
````

**Comment by fragtion** [comment]
My master folder has trash can versioning with "Clean out after" time > 0, so it was pruning empty folders in its .stversions (including my .stfolder). Touching empty .placeholder file in .stfolder solved the problem. Also with expiry time on Zero, it does not delete the empty folder.  Problem solved - tnks!:)

**Comment by milefork** [comment]
A little bump to this, as of today I renamed some folders and now syncthing has created >200 version files and is actually resyncronizing it (over the internet!!!). The hashes are equal!
![image](https://user-images.githubusercontent.com/1383479/62328684-102cdf80-b4b4-11e9-9156-1968254f4422.png)


I think this was caused due to a too fast rename operation(on side B got deleted and recreated?)

As you can see the file would be there but it will be retransmitted over the net 😒

**Comment by vontrapp** [comment]
Since I can't contribute to the other thread, I'll comment here.

For the workaround of having a stversion folder indexed, the problem is that the .stfolder is *inside* the .stversions and as I understand it what calmh suspects is exactly what happens. The versioning cleans up after itself and the .stfolder is removed as part of the cleaning.

What I do instead is create a parent "stversions" folder and tell each folder I want to be versioned to store it's versions as a *subfolder* of that parent stversions. stversions itself is indexed and has a .stfolder in it, but stversions/folder-a-versions/ is cleaned and maintaned by folder 'a', and there is no .stfolder inside of there that gets deleted. I can also have a single index on stversions/ for all versioned files from all other folders.

Indeed the version folder does not have to be a subdirectory of the folder the versions belong to.

And another thought on indexing the versions as a default action, consider that those are indexed in some separate way or even fully synced. The danger pointed out is that multiple hosts would be managing those version files and probably racing each other in many instances. This is indeed a hard problem. The other concern was versioning the version files - this could be easily dealt with by excluding any files within the configured version dir from triggering any versioning actions when those themselves are changed by a human.

**Comment by vontrapp** [comment]
Oh and I have NEVER had good luck with any rename ever copying from the original on a remote node. If I want that to happen I take it as given that I must copy the file instead of move it, wait for everything to be in sync, and then only remove the original location of the file.

If this bit could be improved it would be much appreciated. There is a config option about keeping temporary files for some time period (default 24h iirc) and I mistakenly thought that meant that any file the folder new about would be held on to for that time to make sure it wasn't needed to reuse blocks, and then deleting it after that period. It seems instead this only applies to partially transferred files.
